howtos/diffs/lr-schedule.diff

```diff
@@ -0,0 +1,89 @@
+diff --git a/examples/mnist/train.py b/examples/mnist/train.py
+index 5dc74db..11a5e50 100644
+--- a/examples/mnist/train.py
++++ b/examples/mnist/train.py
+@@ -84,6 +84,16 @@ def create_model(key):
+   model = nn.Model(CNN, initial_params)
+   return model
+ 
++def create_triangular_schedule(lr_min, lr_max, epochs):
++  """return a triangular learning rate schedule"""
++  top = epochs // 2
++  def learning_rate_fn(epoch):
++    if epoch < top:
++      lr = lr_min + epoch/top * (lr_max - lr_min)
++    else:
++      lr = lr_max - (epoch/top - 1)*(lr_max - lr_min)
++    return jnp.asarray(lr, dtype=jnp.float32)
++  return learning_rate_fn
+ 
+ def create_optimizer(model, learning_rate, beta):
+   optimizer_def = optim.Momentum(learning_rate=learning_rate, beta=beta)
+@@ -111,7 +121,7 @@ def compute_metrics(logits, labels):
+ 
+ 
+ @jax.jit
+-def train_step(optimizer, batch):
++def train_step(optimizer, batch, lr):
+   """Train for a single step."""
+   def loss_fn(model):
+     logits = model(batch['image'])
+@@ -119,7 +129,7 @@ def train_step(optimizer, batch):
+     return loss, logits
+   grad_fn = jax.value_and_grad(loss_fn, has_aux=True)
+   (_, logits), grad = grad_fn(optimizer.target)
+-  optimizer = optimizer.apply_gradient(grad)
++  optimizer = optimizer.apply_gradient(grad, learning_rate=lr)
+   metrics = compute_metrics(logits, batch['label'])
+   return optimizer, metrics
+ 
+@@ -130,18 +140,18 @@ def eval_step(model, batch):
+   return compute_metrics(logits, batch['label'])
+ 
+ 
+-def train_epoch(optimizer, train_ds, batch_size, epoch, rng):
++def train_epoch(optimizer, train_ds, batch_size, epoch, rng, learning_rate_fn):
+   """Train for a single epoch."""
+   train_ds_size = len(train_ds['image'])
+   steps_per_epoch = train_ds_size // batch_size
+-
```

Comment:
Why did you remove the whitespace here?

Reply:
added it back after the calling the lr schedule

Reply:
I think the problem is that the whitespace you added below has a training space. If you make it actually an empty line then the diff will be cleaner.